scalableminds/webknossos
# Allow Users to Join Multiple Organizations

Users can now join multiple organizations, admins can now invite users by email address, skipping the manual user activation step.

### TODO Backend
 - [x] move email calls to multiuser
 - [x] only superusers should be able to change emails if multiuser has more than one identity
 - [x] superuser switch route (by email) needs multiuser disambiguation
 - [x] new user switch route by orga
 - [x] organization read access query: allow multiuser indirection
 - [x] set lastLoggedInIdentity upon login and upon switch
 - [x] use lastLoggedInIdentity upon login
 - [x] superusers should see all orgas
 - [x] forbid switching to deactivated users
 - [x] user switch route should lazily create user for superusers
 - [x] invite table in schema, invite case class + service + DAO
 - [x] create + save invite tokens
 - [x] add route to get organization name for invite token
 - [x] assert valid invite tokens in join request
 - [x] invalidate invite tokens after they are used to join
 - [x] also allow *registering* by invite (make sure logic doesn’t break single-orga behavior)
 - [x] send new-user notification upon user join to orga admin
 - [x] send invite emails (remove link from stdout)
 - [x] double-check that change password still works (both reset and change while logged in)
 - [x] repair createOrganizationWithAdmin request (create multi-user, handle loginInfo correctly)
 - [x] add sql constraint for users that (_multiUser, _organization) should be unique
 - [x] make GET routes POST where appropriate
 - [x] db migration
   - [x] adapt schema
   - [x] insert multiusers for each user
   - [x] connect the multiuser references
   - [x] migrate tokens (replace user email with user id)
   - [x] reversion?
 - [x] adapt e2e-test db
 - [x] re-enable unused-warnings in build.sbt
 - [x] switch back `isDemoInstance` default in application.conf to `false`
 - [x] switch back `email.logToStdout` default in application.conf to `false`

### TODO Frontend
 - [x] in user menu (top right of screen), offer to switch to other organizations
   - [x] also make it a bit more obvious that there are multiple organizations
 - [x] change invite UI, no more link, instead, enter recipients, add autoActivate checkbox (default false), send to backend
 - [x] new view `/invite/<tokenValue>`
   - [x] if logged in, show button to join organization with currently active account.
   - [x] if logged out, enforce authentication (either, via registration which should send the `inviteToken` as additional form field) or by logging in (should do an explicit join afterwards)

### New Routes
 - `GET api/auth/joinOrganization/:inviteToken` (allows the user to switchOrganization afterwards)
 - `GET api/auth/switchOrganization/:organizationName` (exchanges user cookie, rediercts to dashboard)
 - `GET api/organizations` (list all you can switch to)
 - `GET api/organizations/byInvite/:inviteToken` (list the organization you have been invited to)
 - `POST api/auth/sendInvites` expects something like `{recipients: ["a@b.com”, "x@y.com"], autoActivate: false}`, for development now logs the invite token to stdout

### Changed Routes
 - `api/auth/register` has new optional form field `inviteToken` (it is required for instances with >1 orga). It uses `organizationName` only if there is no token (to support legacy invite links for the moment) and does not use `organizationDisplayName` anymore. (Those are still used in `api/auth/createOrganizationWithAdmin` though)

### Steps to test:
Hint: set `mail.logToStdout = true` in application.conf to see email contents.

- as demo instance (`features.isDemoInstance = true`) 
  -  create a few organizations
  - invite user 
  - as logged-in user of another org, click invite link, join
  - as logged-in user in that org, click invite link, try to join, should warn that you are already in that org
  - as logged-out user go to invite link, should offer option to register, or log in and then join
  - register via invite link should get you into the matching organization
  - as a user with multiple orgs, use switch menu
  - as superuser, use switch menu, should see all orgs, should be able to switch to all orgs
- as single-org instance
  - invite user
  - register upon invite should land you in that org
  - register without invite should also land you in that org

### Issues:
- fixes #4761 

------
